### PR TITLE
New Permutations() method

### PR DESCRIPTION
**Pull request by varocarbas**
This PM is mostly focused on the algorithm. That is: an efficient, clear and scalable code delivering all the permutations (understood in a wide sense, with up to 4 different configurations). 

As a first step (later it might be extended to other situations), it seems that it should be one of the Array methods (similarly to Array.BinarySearch). Nevertheless, the Array class is in the CoreCLR repository (and this new feature has to be included in CoreFX); that's why I have created a new temporary class.

### CODE REVIEW

**Comment by JonHanna** [comment]
This is an API change. Has it been reviewed? https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md

File: src/Common/src/System/Permutations.cs [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        //That's why a permutation including the positions 0-1-2 (i.e., values 1-2-2) is assumed to not have repetitions.
+        public enum RepetitionsPermutation 
+        { 
+            NoRepetitions,
+            AllRepetitions  
+        };
+        public enum OrderPermutation 
+        { 
+            AnyOrderSamePermutation, //1-2-3 is assumed to be identical to: 1-3-2, 2-3-1, 3-2-1 and 3-1-2.
+            OneOrderOnePermutation   //All the aforementioned permutations are assumed to be different.
```
Comment (JonHanna) [inline src/Common/src/System/Permutations.cs]: `Combination` and `Permutation`?

File: src/Common/src/System/Permutations.cs [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        {
+            return Permutations(Items, Size, RepetitionsPermutation.NoRepetitions, Order);
+        }
+
+        public static Dictionary<int, object[]> Permutations(object[] Items, int Size, RepetitionsPermutation Repetitions, OrderPermutation Order)
+        {
+            return getPermutations(Items, Size, Repetitions, Order);
+        }
+
+        private static Dictionary<int, object[]> getPermutations(object[] Items, int Size, RepetitionsPermutation Repetitions, OrderPermutation Order)
```
Comment (JonHanna) [inline src/Common/src/System/Permutations.cs]: Incorrect casing on method name.

Comment (JonHanna) [inline src/Common/src/System/Permutations.cs]: Incorrect casing on parameters too.

**Comment by varocarbas** [comment]
@JonHanna I created an issue (https://github.com/dotnet/corefx/issues/6162), which BTW I will slightly update in the next minutes.

There was no discussion. But I had a code with the basic structure and I wanted to make my first contribution here, so I started working on it. I reach a pretty good version (in my opinion) and thought that the PR was a good idea. It is always better having a working code than just a bunch of abstract ideas, don't you think?

**Comment by varocarbas** [comment]
@JonHanna with "Incorrect casing on method name." you mean getPermutations I guess. I didn't know that there were casing rules to be applied no matter what, sorry about that. All the public methods are upper case, but not the private ones. I will correct this right away.

**Comment by varocarbas** [comment]
@JonHanna Regarding your "Combination and Permutation?", I mean permutations understood in the widest sense of the expression. I included the following descriptive example in the associated issue:

//All permutations of 2
//a) No repeated elements and order doesn't matter (1-2 = 2-1)
//1-2, 1-3, 2-3
//b) No repeated elements and order matters (1-2 != 2-1) 
//1-2, 1-3, 2-1, 2-3, 3-1, 3-2
//c) Repeated elements and order doesn't matter
//1-1, 1-2, 1-3, 2-3, 2-2, 3-3
//d) Repeated elements and order matters 
//1-1, 1-2, 1-3, 2-1, 2-3, 2-2, 3-1, 3-2, 3-3

**Comment by JonHanna** [comment]
I mean in terms of the name chosen, one of the cases seems to match what's called a combination in mathematics, so maybe the names in the enum could be based on that permutation/combination distinction.

**Comment by stephentoub** [comment]
> It is always better having a working code than just a bunch of abstract ideas, don't you think?

Thanks for your interest in contributing, @varocarbas. As @JonHanna noted, we need to follow the process for new APIs, and until such APIs are actually agreed on for addition, this would be dead code in the repo, which we're trying to avoid. You're welcome to do and evolve such work in your own branch, which you can share a link to in the issue, but a PR at this point won't be merged, so I'm going to close it. Thanks.

**Comment by varocarbas** [comment]
@JonHanna I preferred permutation because combination seems much more abstract (anything can fit there), this is my impression at least. 

I don't think that there is a proper mathematical name for the 4 proposed cases. They are basically pure permutations and some practical (i.e., proven to be required in quite a few scenarios, at least in my experience) extensions.

**Comment by varocarbas** [comment]
@stephentoub I am really sorry to hear that. I was following the expected proceeding, but as far as I had already a worthy code I thought that sharing it was the best option. In fact, the main responsible for this a-bit-sudden PR has been having a quite nice code in place. 

I wasn't expecting it to be merged right away, but being gradually extended/improved with other contributions. I thought that this would be better than just waiting for the go-ahead and creating everything right away.

Anyway... I have to accept your decision. Hopefully, all this will be recovered at some point because I think that it is a quite interesting functionality.

**Pull request closed without merging**